### ISSUE TITLE
(aws_cdk): Aspects applied to App should be inherited by the Stage

### ISSUE THREAD

**Issue by rittneje** [aws/aws-cdk#20468]
### Describe the feature

If I apply an aspect to an App, then all Stages (and all resources within them) should also get the aspect applied, as per usual.

According to @rix0rrr the current behavior is "by design", but I feel this only means the design is faulty.

### Use Case

There are cases where we want to apply tags and/or custom validation aspects to everything in the whole App. We shouldn't have to manually apply them to each Stage. Plus it is already the case that applying an Aspect to the App will recursively apply it to all the Stacks directly in the App, so the current behavior is extremely confusing and unexpected.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [x] This feature might incur a breaking change

It is technically a breaking change, since if today someone is applying an Aspect to the App, it would now start applying to their Stages.

### CDK version used

2.21.1 (build a6ee543)

### Environment details (OS name and version, etc.)

Alpine 3.15, Python 3.9.12

**Comment by rix0rrr** [comment]
Unfortunately Aspects mess with the life cycle of the construct tree. CDK Pipelines needs to be able to look at a part of the construct tree, and be able to determine what stacks are in there and what their dependencies are. It needs to be able to do this during the construction phase, in order to build the right pipeline structure.

If we allowed Aspects to come by later and mutate that state (by adding references or even stacks) then the pipeline structure is incorrect, and people will be caught off guard by that.

Stages are by design a form of isolation of the construct tree... and that requires that they are independent of parent aspects.

**Comment by rittneje** [comment]
@rix0rrr It sounds like the concern is just that an Aspect could add stacks to a Stage and screw up CDK Pipelines. However, you could already do this today by applying an Aspect to the Stage whose `visit` method adds a Stack like so:
```python
def visit(self, c):
   if isinstance(c, aws_cdk.Stack):
      aws_cdk.Stack(c.node.parent, 'another-stack', ...)
```

In other words, Aspects already allow mutating the construct tree in arbitrary ways.

I'm not sure exactly how CDK Pipelines is figuring out the dependency graph today. For reference, we wrote a custom Aspect to do it for similar reasons. (We don't use CDK Pipelines.) So I think the only concept that is missing is how to force such an Aspect to happen at the end after all mutations have been made. One way is to draw a distinction between "mutating" Aspects and "validating" Aspects like Kubernetes does for admission hooks. Another even more flexible way would just to be allow Aspects to have a priority, and then CDK applies Aspects in priority order. This would help in many situations, not just CDK Pipelines.

> Stages are by design a form of isolation of the construct tree... and that requires that they are independent of parent aspects.

This line of reasoning doesn't make sense to me. It's one thing to say that Stages are isolated from each other (i.e., no cross-imports). But it's a pretty big leap from there to claiming they are isolated from their App.

**Comment by rix0rrr** [comment]
> you could already do this today by applying an Aspect to the Stage whose visit method adds a Stack like so:

Yes, but the Stage will apply aspects and `synth` itself before the containing app will do so.

> I'm not sure exactly how CDK Pipelines is figuring out the dependency graph today. 

Like this:

https://github.com/aws/aws-cdk/blob/75bfce70dbc57fe688c96b3c5cbb67fc4e6fcc56/packages/%40aws-cdk/pipelines/lib/blueprint/stage-deployment.ts#L54

This is what `Stage`s were intended for.

> One way is to draw a distinction between "mutating" Aspects and "validating" Aspects like Kubernetes does for admission hooks.

I was thinking something similar. For now, this distinction doesn't exist so we can't rely on it.

> But it's a pretty big leap from there to claiming they are isolated from their App.

Let me rephrase. They are isolated from effects that could require lazy mutation.

**Comment by rittneje** [comment]
@rix0rrr Why can't whatever CDK Pipelines is doing today just happen after Aspects get applied? And why/how is a Stage getting synthed before the App that contains it?

**Comment by peterwoodworth** [comment]
Closing because this is marked as `wontfix` and it's not a bug
